File: apps/java/src/test/java/chapter15/PredictionApiTest.java
package chapter15;

import static org.assertj.core.api.Assertions.assertThat;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import io.javalin.testtools.JavalinTest;
import io.javalin.testtools.Response;
import java.util.Map;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

class PredictionApiTest {
  private static final ObjectMapper MAPPER = new ObjectMapper();
  private static final ModelStore STUB = Stubs.store(movie -> 1200.0, passenger -> true);

  private static ObjectNode movieJson() {
    return MAPPER
        .createObjectNode()
        .put("sns1", 200.0)
        .put("sns2", 500.0)
        .put("actor", 3000.0)
        .put("original", 1);
  }

  private static ObjectNode passengerJson() {
    ObjectNode node =
        MAPPER.createObjectNode().put("pclass", 1).put("sex", "female").putNull("age");
    return node.put("sib_sp", 0).put("parch", 0).put("fare", 50.0).put("embarked", "C");
  }

  private static JsonNode json(Response response) throws Exception {
    return MAPPER.readTree(response.body().string());
  }

  private static JsonNode json(Object value) {
    return MAPPER.valueToTree(value);
  }

  @Nested
  class CinemaSales {
    @Test
    @DisplayName("映画の特徴量を送ると予測した興行収入を返す")
    void predicts() {
      JavalinTest.test(
          PredictionApi.create(new PredictionService(STUB)),
          (server, client) -> {
            Response response = client.post("/cinema/sales", movieJson().toString());

            assertThat(response.code()).isEqualTo(200);
            assertThat(json(response)).isEqualTo(json(Map.of("sales", 1200.0)));
          });
    }

    @ParameterizedTest(name = "{0}={1}")
    @CsvSource(
        delimiter = '|',
        value = {"sns1|-1.0", "actor|\"多い\"", "original|2", "sns2|null"})
    @DisplayName("特徴量が不正なら 422 を返す")
    void rejectsInvalid(String field, String value) throws Exception {
      ObjectNode body = movieJson();
      body.set(field, MAPPER.readTree(value));
      JavalinTest.test(
          PredictionApi.create(new PredictionService(STUB)),
          (server, client) ->
              assertThat(client.post("/cinema/sales", body.toString()).code()).isEqualTo(422));
    }

    @Test
    @DisplayName("モデルが無ければ 503 を返す")
    void missingModel() {
      JavalinTest.test(
          PredictionApi.create(new PredictionService(Stubs.emptyStore())),
          (server, client) -> {
            Response response = client.post("/cinema/sales", movieJson().toString());

            assertThat(response.code()).isEqualTo(503);
            assertThat(json(response)).isEqualTo(json(Map.of("detail", "学習済みモデル cinema が見つかりません")));
          });
    }
  }

  @Nested
  class Survived {
    @Test
    @DisplayName("乗客の特徴量を送ると生存の予測を返す")
    void predicts() {
      JavalinTest.test(
          PredictionApi.create(new PredictionService(STUB)),
          (server, client) -> {
            Response response = client.post("/survived", passengerJson().toString());

            assertThat(response.code()).isEqualTo(200);
            assertThat(json(response)).isEqualTo(json(Map.of("survived", true)));
          });
    }

    @Test
    @DisplayName("年齢と乗船港は省略できる")
    void optionalFields() {
      ObjectNode body = passengerJson();
      body.remove("age");
      body.remove("embarked");
      JavalinTest.test(
          PredictionApi.create(new PredictionService(STUB)),
          (server, client) ->
              assertThat(client.post("/survived", body.toString()).code()).isEqualTo(200));
    }

    @ParameterizedTest(name = "{0}={1}")
    @CsvSource(
        delimiter = '|',
        value = {"pclass|4", "sex|\"unknown\"", "embarked|\"X\"", "fare|-5.0"})
    @DisplayName("特徴量が不正なら 422 を返す")
    void rejectsInvalid(String field, String value) throws Exception {
      ObjectNode body = passengerJson();
      body.set(field, MAPPER.readTree(value));
      JavalinTest.test(
          PredictionApi.create(new PredictionService(STUB)),
          (server, client) ->
              assertThat(client.post("/survived", body.toString()).code()).isEqualTo(422));
    }

    @Test
    @DisplayName("JSON として読めなければ 422 と、内部の型名を含まない理由を返す")
    void rejectsBrokenJson() {
      JavalinTest.test(
          PredictionApi.create(new PredictionService(STUB)),
          (server, client) -> {
            Response response = client.post("/survived", "{");

            assertThat(response.code()).isEqualTo(422);
            assertThat(json(response))
                .isEqualTo(json(Map.of("detail", java.util.List.of("JSON の形式または値の型が正しくありません"))));
          });
    }

    @Test
    @DisplayName("モデルが無ければ 503 を返す")
    void missingModel() {
      JavalinTest.test(
          PredictionApi.create(new PredictionService(Stubs.emptyStore())),
          (server, client) ->
              assertThat(client.post("/survived", passengerJson().toString()).code())
                  .isEqualTo(503));
    }
  }

  @Nested
  class Health {
    @Test
    @DisplayName("すべてのモデルを読み込めれば ok を返す")
    void ok() {
      JavalinTest.test(
          PredictionApi.create(new PredictionService(STUB)),
          (server, client) -> {
            Response response = client.get("/health");

            assertThat(response.code()).isEqualTo(200);
            assertThat(json(response))
                .isEqualTo(
                    json(
                        Map.of(
                            "status", "ok", "models", Map.of("cinema", true, "survived", true))));
          });
    }

    @Test
    @DisplayName("読み込めないモデルがあれば degraded を返す")
    void degraded() {
      JavalinTest.test(
          PredictionApi.create(new PredictionService(Stubs.emptyStore())),
          (server, client) ->
              assertThat(json(client.get("/health")))
                  .isEqualTo(
                      json(
                          Map.of(
                              "status",
                              "degraded",
                              "models",
                              Map.of("cinema", false, "survived", false)))));
    }
  }
}

File: apps/dotnet/tests/MachineLearning.Tests/Chapter15/PredictionApiTest.fs
module MachineLearning.Tests.Chapter15.PredictionApiTest

open System.Net
open System.Net.Http
open System.Text
open System.Threading.Tasks
open Microsoft.AspNetCore.Builder
open Microsoft.AspNetCore.TestHost
open Xunit
open MachineLearning.Chapter15.Domain
open MachineLearning.Chapter15.PredictionApi
open MachineLearning.Tests.Chapter15.Stubs

/// 置き場を使う API を、ネットワークを使わないテスト用のサーバーで起動し、クライアントを返す
let clientWith (store: ModelStore) : Task<HttpClient> =
    task {
        let builder = WebApplication.CreateBuilder()
        builder.WebHost.UseTestServer() |> ignore
        let app = createWebApplication builder store
        do! app.StartAsync()
        return app.GetTestClient()
    }

/// 要求を送り、状態コードと本文の組を返す
let send (store: ModelStore) (request: HttpRequestMessage) : Task<HttpStatusCode * string> =
    task {
        let! client = clientWith store
        use! response = client.SendAsync request
        let! body = response.Content.ReadAsStringAsync()
        return response.StatusCode, body
    }

let postJson (store: ModelStore) (path: string) (body: string) : Task<HttpStatusCode * string> =
    let request = new HttpRequestMessage(HttpMethod.Post, path)
    request.Content <- new StringContent(body, Encoding.UTF8, "application/json")
    send store request

let get (store: ModelStore) (path: string) : Task<HttpStatusCode * string> =
    send store (new HttpRequestMessage(HttpMethod.Get, path))

let movieJson = """{"sns1": 200, "sns2": 500, "actor": 3000, "original": 1}"""

let passengerJson =
    """{"pclass": 1, "sex": "female", "age": null, "sib_sp": 0, "parch": 0, "fare": 50, "embarked": "C"}"""

[<Fact>]
let ``映画の特徴量を送ると予測した興行収入を返す`` () =
    task {
        let! actual = postJson stubModelStore "/cinema/sales" movieJson
        Assert.Equal((HttpStatusCode.OK, """{"sales":1200}"""), actual)
    }

[<Fact>]
let ``特徴量が不正なら 422 と理由を返す`` () =
    task {
        let body = """{"sns1": -1, "sns2": 500, "actor": 3000, "original": 1}"""
        let! actual = postJson stubModelStore "/cinema/sales" body
        Assert.Equal((HttpStatusCode.UnprocessableEntity, """{"detail":["sns1 は 0 以上にしてください"]}"""), actual)
    }

[<Fact>]
let ``JSON として読めなければ 422 を返す`` () =
    task {
        let! actual = postJson stubModelStore "/cinema/sales" "{"
        Assert.Equal((HttpStatusCode.UnprocessableEntity, """{"detail":["JSON の形式が正しくありません"]}"""), actual)
    }

[<Fact>]
let ``興行収入のモデルが無ければ 503 を返す`` () =
    task {
        let! actual = postJson emptyModelStore "/cinema/sales" movieJson
        Assert.Equal((HttpStatusCode.ServiceUnavailable, """{"detail":"学習済みモデル cinema が見つかりません"}"""), actual)
    }

[<Fact>]
let ``乗客の特徴量を送ると生存の予測を返す`` () =
    task {
        let! actual = postJson stubModelStore "/survived" passengerJson
        Assert.Equal((HttpStatusCode.OK, """{"survived":true}"""), actual)
    }

[<Fact>]
let ``生存のモデルが無ければ 503 を返す`` () =
    task {
        let! actual = postJson emptyModelStore "/survived" passengerJson
        Assert.Equal(HttpStatusCode.ServiceUnavailable, fst actual)
    }

[<Fact>]
let ``すべてのモデルを読み込めれば ok を返す`` () =
    task {
        let! actual = get stubModelStore "/health"
        Assert.Equal((HttpStatusCode.OK, """{"status":"ok","models":{"cinema":true,"survived":true}}"""), actual)
    }

[<Fact>]
let ``読み込めないモデルがあれば degraded を返す`` () =
    task {
        let! actual = get emptyModelStore "/health"

        Assert.Equal(
            (HttpStatusCode.OK, """{"status":"degraded","models":{"cinema":false,"survived":false}}"""),
            actual
        )
    }

[<Fact>]
let ``予測中に例外が起きたら内部の情報を出さずに 500 を返す`` () =
    task {
        let failingStore =
            { stubModelStore with
                LoadSalesModel = fun () -> Ok(fun _ -> failwith "内部の詳細")
            }

        let! actual = postJson failingStore "/cinema/sales" movieJson
        Assert.Equal((HttpStatusCode.InternalServerError, """{"detail":"予測中にエラーが発生しました"}"""), actual)
    }

[<Fact>]
let ``無い経路には 404 を返す`` () =
    task {
        let! actual = get stubModelStore "/unknown"
        Assert.Equal(HttpStatusCode.NotFound, fst actual)
    }
